Move Alipay and Multibanco inside Stripe settings when UPE is enabled

Mayisha opened the pull request:
Fixes #2852 

## Changes proposed in this Pull Request:
- Alipay and Multibanco are moved inside the Stripe settings page for UPE enabled state.

**Before**

![image](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/33387139/3f6fe192-e5e6-4b90-ba07-807f4c4bbb54)

**After**
<img src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/33387139/66e2e0b8-8779-4a80-831f-ed50d8d27b3b" width="50%" height="50%" />

- On checkout page, Alipay is now rendered as an UPE method. As Multibanco does not use Payments API of Stripe (uses the Stripe Sources API), it can not be rendered as an UPE method. 

**Alipay on checkout page**
<img src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/33387139/ec60fda6-cfb4-46de-b553-ffb49ffee8ac" width="50%" height="50%" />

**Multibanco on checkout page**
<img src="https://github.com/woocommerce/woocommerce-gateway-stripe/assets/33387139/8c5cffb8-124b-4af4-806e-d6111c4771bf" width="50%" height="50%" />


## Testing instructions
- Go to the Stripe settings page and enable UPE.
- Go to the Payment Methods tab on the Stripe settings page. Alipay and Multibanco should be listed here. ( You may not see them on the list as their capabilities are not set by default as per the document https://stripe.com/docs/connect/account-capabilities#payment-methods. You have to activate them from the Stripe dashboard. As I don't have an activated account and could not activate Alipay and Multibanco I tweaked the capabilities [here](https://github.com/woocommerce/woocommerce-gateway-stripe/blob/892d8a3d7bc2f7a6a705cd0701295e4c4a411b51/client/settings/general-settings-section/payment-methods-list.js#L170) to make them appear in the list).
- Enable/Disable these 2 methods and make sure they are available on the checkout page as per settings.
- Alipay should be rendered as an UPE method, Multibanco should be rendered as a separate payment method as before.
- Purchasing with Alipay and Multibanco should work without errors.

### Note (ToDo)
Multibanco has been moved inside the same Stripe settings page when UPE is enabled but it can not be used as an UPE method at this moment because of using Sources API. When Stripe starts supporting Multibanco with Payments API and it can be moved inside as an UPE method, we should revert the Multibanco specific changes of this PR and create a new UPE class for Multibanco.

Mayisha committed:
move alipay and multibanco inside stripe settings page

Mayisha committed:
show alipay and multibanco icons beside their names

Mayisha committed:
handle edge case

Mayisha committed:
Merge branch 'develop' into task/2852-missing-non-upe-methods-in-settings

Mayisha committed:
move alipay and multibanco only when upe is disabled

Mayisha committed:
fix tests

Mayisha committed:
use helper

Mayisha committed:
add Alipay as a upe method

Mayisha committed:
show multibanco on the stripe settings page

Mayisha committed:
enable/disable multibanco from the settings page

Mayisha committed:
show multibanco on the checkout page

Mayisha committed:
Merge branch 'develop' into task/2852-missing-methods-in-settings-for-upe

Mayisha committed:
show alipay as disabled correctly when upe is enabled

Mayisha committed:
remove unused files

Mayisha committed:
update tests

Mayisha committed:
Merge branch 'develop' into task/2852-missing-methods-in-settings-for-upe

Mayisha commented on includes/admin/class-wc-rest-stripe-payment-gateway-controller.php:
```diff
@@ -1,175 +0,0 @@
-		$value = sanitize_text_field( $description );
-		$this->gateway->update_option( 'description', $value );
-	}
-}
```
As all the methods are moved inside Stripe settings page, this separate page for the individual methods is not needed anymore. 

![separate](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/33387139/f35de1f9-b2d6-4f33-9e6e-0cb58816e42b)


So deleting the controller and relevant tests. Also the tests started to fail as we are not registering any payment method for the payment settings page.

Mayisha commented on includes/class-wc-stripe-helper.php:
```diff
@@ -545,6 +545,41 @@ public static function get_legacy_individual_payment_method_settings() {
+
+		return $enabled_gateways;
+	}
+
```
These helper methods are temporarily added to include Multibanco to the list so that it appears on the Stripe settings page. When Stripe starts supporting Multibanco and we create a new UPE class for Multibanco, these methods won't be needed anymore.

mattallan replied:
Makes sense! Thanks for the explanation :)

Mayisha commented on includes/payment-methods/class-wc-stripe-upe-payment-method-alipay.php:
```diff
@@ -0,0 +1,94 @@
+	 *
+	 * @return array
+	 */
+	public function get_supported_currencies() {
```
As I moved Alipay inside UPE, I was getting an error on the checkout page when I was using an `US` Stripe account with store currency set as `EURO`. Also the UPE payment methods failed to render on the page because of this error.

![error alipay](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/33387139/1184a74c-6f93-47ff-9796-1df4dda939a6)
 
As per my understanding after reading https://stripe.com/docs/payments/alipay#supported-currencies that says `You can create Alipay payments in the currencies that map to your country` I have created this mapping function that returns the supported currency list based on the Stripe account's country. 

Now on the settings page also when UPE is enabled, if your Stripe account is a US account but WC store currency is Euro, it will show Alipay as unsupported for currency.

To the reviewer: please check if my assumption is correct here.

mattallan commented on includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:
```diff
@@ -395,6 +424,7 @@ public function get_upe_enabled_payment_method_ids() {
 	public function get_upe_enabled_at_checkout_payment_method_ids( $order_id = null ) {
 		$is_automatic_capture_enabled = $this->is_automatic_capture_enabled();
 		$available_method_ids         = [];
+		WC_Stripe_Logger::log( 'get_upe_enabled_at_checkout_payment_method_ids---' );
```
Just checking if this code is intentional, or is this debugging code that accidentally made it into the PR? 😅

There's a few more instances of WC_Stripe_Logger::log() used below this one as well.

Mayisha replied:
:woman_facepalming:

mattallan commented on client/payment-methods-map.js:
```diff
@@ -12,6 +12,92 @@ import OxxoIcon from './payment-method-icons/oxxo';
+	];
+
+	// cached_account_data = WC_Stripe::get_instance()->account->get_cached_account_data();
+	const country = 'US'; //cached_account_data['country'] ?? null;
```
I'm assuming this was hardcoded to 'US' for testing?

Mayisha replied:
Yup this was for testing. I will fix it :sweat_smile:

mattallan reviewed (commented):
Hey Mayisha! I've left some minor feedback from code review but I've been having trouble getting Alipay to show up in the payment method list.

When I enable UPE this is what I see:
![image](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/2275145/e42d2637-b436-4ff7-954b-d0c49adc316f)

From my Stripe dashboard, I can see that Alipay is active/enabled: 
![image](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/2275145/292d0c40-ac55-49e1-a5d5-017731376071)

When I disable UPE I can see Alipay: ![image](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/2275145/e22f0693-9bde-4b75-aac3-1a39fdb73cf3)

I noticed you said you had to modify the capabilities, but since I have Alipay enabled, I wanted to double-check there's not a bug that is causing this 🤔

Mayisha committed:
Merge branch 'develop' into task/2852-missing-methods-in-settings-for-upe

Mayisha committed:
remove logs added for debugging

Mayisha committed:
Address feedback
- check capabilities only for live mode and UPE enabled
- create hook for evaluating Alipay currencies

Mayisha committed:
fix test

Mayisha committed:
fix list items issues
- fix payment methods after sofort being discarded because of not adding number of elements in `splice`
- fix logic to eleminate link from the list correctly

Mayisha committed:
Merge branch 'develop' into task/2852-missing-methods-in-settings-for-upe

Mayisha committed:
refactor payment method list logic

mattallan commented on client/utils/use-alipay-currencies.js:
```diff
@@ -0,0 +1,93 @@
+	}
+
+	switch ( data?.account?.country ) {
+		case 'AUS':
```
I couldn't enable AliPay on my stripe account even though it had a supported currency (i.e. AUD) and I found out the issue is with this mapping function. My AUS Stripe account has the data.account.country value set to `AU` not `AUS`, I also connected a Canadian account and the data.account.country was `CA`, not `Canada`.

Has something changed since we populated this switch statement and do we need to update these switch cases to use the country codes instead?

From my testing the following account.country values need to be updated:
- `AUS` -> `AU`
- `Canada` -> `CA`
- `HongKong` -> `HK`
- `Japan` -> `JP`
- `Malaysia` -> `MY`
- `Singapore` -> `SG`

Mayisha replied:
It was a silly mistake on my end. I used country code for some and full country name for some cases :woman_facepalming: 

Thanks @mattallan. Fixed in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2870/commits/686f6c8d119b0f87e2a9a364eb9ef3c5e85ebb2e

Mayisha committed:
fix country names

Mayisha committed:
Merge branch 'develop' into task/2852-missing-methods-in-settings-for-upe

mattallan commented on includes/payment-methods/class-wc-stripe-upe-payment-gateway.php:
```diff
@@ -166,6 +167,34 @@ public function __construct() {
+		$gateways_upto_stripe  = array_slice( $gateways, 0, $stripe_index + 1 );
+		$gateways_after_stripe = array_slice( $gateways, $stripe_index + 1 );
+
+		$multibanco_gateway = new WC_Gateway_Stripe_Multibanco();
```
Instantiating a new instance of the Multibanco class here is causing all of the hooks inside the __construct to be registered multiple times, causing things like this on the thank-you page and probably other issues: 
![image](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/2275145/19284c41-d068-46d2-99cb-9274beda84f0)

mattallan reviewed (changes requested):
Thanks @Mayisha for all of your work on this PR!! 💯 

- [x] Alipay and Multibanco now part of the payment methods list within Stripe Settings.
- [x] Confirmed Alipay now part of the UPE on checkout
  - [x] tested purchasing with Alipay on both shortcode and block checkouts 👍 
- [x] Confirmed Multibanco is not part of the UPE
  - [ ] tested purchasing with multibanco checkout :x:

I'm noticing a couple of issues while testing Multibanco when comparing things to `develop`.

- On Edit Order page the "Payment via " is display payment method instead of the expected title

I had a quick look into this and couldn't find the cause. I'm thinking maybe it has something to do with the multibanco class missing a `get_title()` function?

| `develop` | this branch |
|--------|--------|
|![image](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/2275145/1398bac3-b869-4b78-a3e6-270b8b443962)|![image](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/2275145/6968850f-af1a-479d-a090-5c71af8b1430)| 

- Order thankyou page instructions are being duplicated

This is being caused by multiple instances of `WC_Gateway_Stripe_Multibanco` being instantiated. I left a comment below with which code I think it's coming from.

| `develop` | this branch |
|--------|--------|
|![image](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/2275145/4aad2b55-822b-4039-bb0c-cd7f79451ab3)|![image](https://github.com/woocommerce/woocommerce-gateway-stripe/assets/2275145/866ea185-73b9-442e-a922-336fd939bb30)|

Mayisha committed:
use multibanco instance from helper class via get_legacy_payment_method

Mayisha commented:
Thanks @mattallan for the thorough review and for finding the issues. :+1: 

> On Edit Order page the "Payment via " is display payment method instead of the expected title

As we found this issue to be present in `develop` and also for non-UPE settings, I would like to handle this as a separate issue.

> Order thankyou page instructions are being duplicated

I have removed the Multibanco instantiation in https://github.com/woocommerce/woocommerce-gateway-stripe/pull/2870/commits/ddc8d7052041fff48f0bf0ba03bf98e426988b0c and now using the instance from the helper method.

Mayisha committed:
do not load multibanco for subscription product on checkout page

Mayisha committed:
move check in multibanco class

mattallan reviewed (approved):
Thanks @Mayisha! I've tested the latest changes and didn't find any issues! 💯
Code also looks good to me :)

Time to ship it!!
